### ISSUE TITLE
Pyright thinks code after `while True` is unreachable although it might be

### ISSUE THREAD

**Issue by cleoold** [microsoft/pyright#1431]
**Describe the bug**

**To Reproduce**
let pyright check the following code, it will warn `self.cleanup()` and say:
```
Cannot access member "cleanup" for type "Never"
  Type "Never" is unsupported
```

**Expected behavior**
It should somehow allow me to do this because this code is reachable and there is a `cleanup` method defined.

**Screenshots or Code**
```python
from typing import Any
import asyncio

class Myjob:
    async def do_stuff(self) -> Any: # inferred as NoReturn
        try:
            while True:
                await asyncio.sleep(1)
                print('hello')
        finally:
            self.cleanup() # self is inferred as Never so I cannot call .cleanup() on this variable

    def cleanup(self):
        print('bye')

async def main():
    c = asyncio.create_task(Myjob().do_stuff())
    await asyncio.sleep(5)
    c.cancel()

asyncio.run(main())
```

Running this program, it prints four "hello" and one "bye", then exits.

**VS Code extension or command-line**
VS Code extension 1.1.106

**Additional context**
We use the code here: https://github.com/nonebot/aiocqhttp/blob/a87da2968848e5c66eade401cd8ab1051c5ed3e4/aiocqhttp/__init__.py#L531

**Comment by erictraut** [comment]
Thanks for the bug report.

Finally clauses are very difficult to model correctly in a static analyzer because control falls through the bottom of the finally clause if it is invoked without an exception or a return (or sometimes a break or continue). That makes it impossible to draw a static control flow graph that captures all possible control transitions. 

Type checkers like Pyright and the TypeScript compiler use some clever tricks to approximate a control flow graph that works in most cases. You have discovered an edge case where that trick falls apart.

I'll see if I can find some ways to extend the existing trick so as to make this case work. In the meantime, if you're looking for a workaround, adding a dummy except clause will eliminate the error.

```python
        except BaseException as e:
            raise e
```

**Comment by erictraut** [comment]
I've introduced a fix for this. I'm a little worried about the fix because it has the potential to produce false positives in certain cases. But I think it's a better solution that the current approach because there is an obvious way to fix the false positives resulting from the new technique — and arguably make the code better. The false positive that you are seeing in your code has no straightforward workaround.

This will be addressed in the next release.

**Comment by cleoold** [comment]
Thanks for your great effort :-] I am super impressed by your quick and detailed responses to each bugs in issues. Again thanks

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.107, which I just published. It will also be included in the next release of Pylance.
